client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java

```diff
@@ -79,6 +91,36 @@ public class SparkShuffleManager implements ShuffleManager {
   private long sendBufferPoolCheckInterval;
   private long sendBufferPoolExpireTimeout;
 
+  // track appShuffleId -> shuffleId Set in executor for cleanup
+  static class AppShuffleIdTracker {
+    private ConcurrentHashMap<Integer, HashSet<Integer>> shuffleIdMap =
+        JavaUtils.newConcurrentHashMap();
+
+    public void track(int appShuffleId, int shuffleId) {
+      HashSet<Integer> shuffleIds =
+          shuffleIdMap.computeIfAbsent(
+              appShuffleId,
+              new Function<Integer, HashSet<Integer>>() {
+                @Override
+                public HashSet<Integer> apply(Integer id) {
+                  return new HashSet<Integer>(shuffleId);
+                }
+              });
+      synchronized (shuffleIds) {
+        shuffleIds.add(shuffleId);
+      }
+    }
+
+    public void unregisterAppShuffleId(ShuffleClient shuffleClient, int appShuffleId) {
+      HashSet<Integer> shuffleIds = shuffleIdMap.remove(appShuffleId);
+      synchronized (shuffleIds) {
+        shuffleIds.forEach(shuffleId -> shuffleClient.cleanupShuffle(shuffleId));
+      }
```

Comment:
In case of race or concurrent updates
```suggestion
      if (null != shuffleIds) {
        synchronized (shuffleIds) {
          shuffleIds.forEach(shuffleId -> shuffleClient.cleanupShuffle(shuffleId));
        }
      }  
```

Reply:
Agree, this is a good catch!